libnetwork/portmapper/proxy_linux.go

```diff
@@ -1,78 +1,61 @@
 package portmapper
 
 import (
+	"errors"
 	"fmt"
 	"io"
-	"net"
 	"os"
 	"os/exec"
 	"runtime"
 	"strconv"
 	"syscall"
 	"time"
-)
 
-// StartProxy starts the proxy process at proxyPath, or instantiates a dummy proxy
-// to bind the host port if proxyPath is the empty string.
-func StartProxy(
-	proto string,
-	hostIP net.IP, hostPort int,
-	containerIP net.IP, containerPort int,
-	proxyPath string,
-) (stop func() error, retErr error) {
-	if proxyPath == "" {
-		return newDummyProxy(proto, hostIP, hostPort)
-	}
-	return newProxyCommand(proto, hostIP, hostPort, containerIP, containerPort, proxyPath)
-}
+	"github.com/docker/docker/libnetwork/types"
+)
 
-func newProxyCommand(
-	proto string,
-	hostIP net.IP, hostPort int,
-	containerIP net.IP, containerPort int,
+// StartProxy starts the proxy process at proxyPath.
+// If listenSock is not nil, it must be a bound socket that can be passed to
+// the proxy process for it to listen on.
+func StartProxy(pb types.PortBinding,
 	proxyPath string,
+	listenSock *os.File,
 ) (stop func() error, retErr error) {
 	if proxyPath == "" {
 		return nil, fmt.Errorf("no path provided for userland-proxy binary")
 	}
+	r, w, err := os.Pipe()
+	if err != nil {
+		return nil, fmt.Errorf("proxy unable to open os.Pipe %s", err)
+	}
+	defer func() {
+		if w != nil {
+			w.Close()
+		}
+		r.Close()
+	}()
 
-	p := &proxyCommand{
-		cmd: &exec.Cmd{
-			Path: proxyPath,
-			Args: []string{
-				proxyPath,
-				"-proto", proto,
-				"-host-ip", hostIP.String(),
-				"-host-port", strconv.Itoa(hostPort),
-				"-container-ip", containerIP.String(),
-				"-container-port", strconv.Itoa(containerPort),
-			},
-			SysProcAttr: &syscall.SysProcAttr{
-				Pdeathsig: syscall.SIGTERM, // send a sigterm to the proxy if the creating thread in the daemon process dies (https://go.dev/issue/27505)
-			},
+	cmd := &exec.Cmd{
+		Path: proxyPath,
+		Args: []string{
+			proxyPath,
+			"-proto", pb.Proto.String(),
+			"-host-ip", pb.HostIP.String(),
```

Comment:
This has to consume `childIP` chosen by the former `rootlesskit-docker-proxy` code
https://github.com/rootless-containers/rootlesskit/blob/4fb2e2cb80bf13eb28b7f2a4317b63406b89ad32/cmd/rootlesskit-docker-proxy/main.go#L200

Reply:
Ohhh, thank you - I'd missed that, it all makes more sense now!

(I'd only tested with no host address specified, so the proxy bound to all addresses in the child namespace. Now tested with specific host addresses as well, as shown in the description.)

`DNAT` rules now also use the host IP (which makes a port mapping with a specific host address work in rootless mode with `--userland-proxy=false`).